src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -222,24 +224,24 @@ internal static SafeSslHandle AllocateSslHandle(SafeFreeSslCredentials credentia
             SafeSslContextHandle? sslCtxHandle = null;
             SafeSslContextHandle? newCtxHandle = null;
             SslProtocols protocols = CalculateEffectiveProtocols(sslAuthenticationOptions);
+            bool hasAlpn = sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0;
             bool cacheSslContext = !DisableTlsResume && sslAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.RequireEncryption &&
+                    sslAuthenticationOptions.IsServer &&
                     sslAuthenticationOptions.CertificateContext != null &&
                     sslAuthenticationOptions.CertificateContext.SslContexts != null &&
-                    sslAuthenticationOptions.CipherSuitesPolicy == null &&
-                    (!sslAuthenticationOptions.IsServer ||
-                    (sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0));
+                    sslAuthenticationOptions.CipherSuitesPolicy == null;
 
             if (cacheSslContext)
             {
-               sslAuthenticationOptions.CertificateContext!.SslContexts!.TryGetValue(protocols, out sslCtxHandle);
+               sslAuthenticationOptions.CertificateContext!.SslContexts!.TryGetValue(protocols | (SslProtocols)(hasAlpn ? 1 : 0), out sslCtxHandle);
```

Comment:
This seems a bit tenuous... do we have any asserts anywhere or some other protection that will warn us if we ever use the bottom bit of SslProtocols?  Alternatively, you could create a simple struct to use as the key.

Reply:
Since this comes via CalculateEffectiveProtocols() I can add checks/assert/cleanup there.

Reply:
Updated `CalculateEffectiveProtocols` to make sure the low bit is never set. I see no valid use and even if set now, it would be ignored by the PAL.

Reply:
> Updated CalculateEffectiveProtocols to make sure the low bit is never set.

Well, sort of.  It clears the bit.  If the TLS1.4/2.0/whatever identifier had the bit set then the routine would just decide it was irrelevant.

If unknown values are caught before CalculateEffectiveProtocols, then it should instead assert that the bit was clear, instead of clearing it.  If CalculateEffectiveProtocols can be fed garbage then we should add a test that does something like

```C#
[Fact]
public static void SslProtocolsLowBitAvailable()
{
    foreach (SslProtocols value in Enum.GetValues<SslProtocols>())
    {
        Assert.True((value & (SslProtocols)1) == 0, $"SslProtocols.{value} does not use the least significant bit");
    }
}
```

Reply:
I added assert to catch case if low bit is set for valid reason  e.g. we have new enum added that uses it, 
I also look again at the pa again:
https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Security.Cryptography.Native/pal_ssl.c#L248-L269

It will effectively ignore unknown values.